### PR TITLE
Introduce parameterized stake token

### PR DESCRIPTION
**Pull request by dusannosovic-ethernal**
# Description

This PR includes stake-token flag in diffrent commands, now we can choose stake token in stake manager.

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [X] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Checklist

- [X] I have assigned this PR to myself
- [X] I have added at least 1 reviewer
- [ ] I have added the relevant labels
- [ ] I have updated the official documentation
- [ ] I have added sufficient documentation in code

## Testing

- [ ] I have tested this code with the official test suite
- [ ] I have tested this code manually

### CODE REVIEW

File: command/genesis/params.go [hunk 1]
```diff
@@ -187,6 +191,8 @@ func (p *genesisParams) validateFlags() error {
 		}
 	}
 
+	params.stakeTokenAddr = types.StringToAddress(params.stakeToken)
```
Comment (goran-ethernal) [inline command/genesis/params.go]: Add validation of `stakeToken` address, like we have in other commands. (Use `types.IsValidAddress(contractAddress)` function).

File: command/validator/registration/params.go [hunk 2]
```diff
@@ -28,6 +31,8 @@ func (rp *registerParams) validateFlags() (err error) {
 		return fmt.Errorf("provided value (%d) is less than zero", rp.amountValue)
 	}
 
+	params.stakeTokenAddr = types.StringToAddress(params.stakeToken)
```
Comment (goran-ethernal) [inline command/validator/registration/params.go:41]: The same. Add validation of the address.

File: command/validator/stake/params.go [hunk 3]
```diff
@@ -30,6 +33,8 @@ func (sp *stakeParams) validateFlags() (err error) {
 		return fmt.Errorf("failed to parse json rpc address. Error: %w", err)
 	}
 
+	sp.stakeTokenAddr = types.StringToAddress(params.stakeToken)
```
Comment (goran-ethernal) [inline command/validator/stake/params.go]: Same.

File: command/genesis/genesis.go [hunk 4]
```diff
@@ -241,6 +242,13 @@ func setFlags(cmd *cobra.Command) {
 			defaultWithdrawalWaitPeriod,
 			"number of epochs after which withdrawal can be done from child chain",
 		)
+
+		cmd.Flags().StringVar(
+			&params.stakeToken,
+			stakeTokenFlag,
+			contracts.NativeERC20TokenContract.String(),
+			"address of staking token used in stake manager for staking",
```
Comment (Stefan-Ethernal) [inline command/genesis/genesis.go]:
```suggestion
			"stake token address",
```

File: consensus/polybft/polybft_config.go [hunk 5]
```diff
@@ -94,6 +94,8 @@ type PolyBFTConfig struct {
 
 	// GovernanceConfig defines on chain governance configuration
 	GovernanceConfig *GovernanceConfig `json:"governanceConfig"`
+
+	StakeTokenAddr types.Address `json:"stakeTokenAddr"`
```
Comment (Stefan-Ethernal) [inline consensus/polybft/polybft_config.go]:
```suggestion
    // StakeTokenAddr represents the stake token contract address
	StakeTokenAddr types.Address `json:"stakeTokenAddr"`
```

File: command/genesis/params.go [hunk 6]
```diff
@@ -187,6 +192,16 @@ func (p *genesisParams) validateFlags() error {
 		}
 	}
```
Comment (Stefan-Ethernal) [inline command/genesis/params.go]:
```suggestion
    // Validate stake token address
```

File: command/genesis/params.go [hunk 7]
```diff
@@ -187,6 +192,16 @@ func (p *genesisParams) validateFlags() error {
...
 
+	if err := types.IsValidAddress(params.stakeToken); err != nil {
+		return fmt.Errorf("stake token address is not a valid address: %w", err)
+	}
+
+	params.stakeTokenAddr = types.StringToAddress(params.stakeToken)
+
+	if params.stakeTokenAddr == types.ZeroAddress {
+		return errStakeTokenIsZeroAddress
+	}
```
Comment (Stefan-Ethernal) [inline command/genesis/params.go]: Consider moving this validation where we are checking if it is `polybft` consensus (`if p.isPolyBFTConsensus()`).

File: command/secrets/init/utils.go [hunk 8]
```diff
@@ -15,12 +15,14 @@ const (
...
 	ChainIDFlag       = "chain-id"
 	AmountFlag        = "amount"
+	StakeTokenFlag    = "stake-token"
 
 	AccountDirFlagDesc    = "the directory for the Polygon Edge data if the local FS is used"
 	AccountConfigFlagDesc = "the path to the SecretsManager config file, if omitted, the local FS secrets manager is used"
 	PrivateKeyFlagDesc    = "hex-encoded private key of the account which executes command"
 	ChainIDFlagDesc       = "ID of child chain"
 	AmountFlagDesc        = "amount to stake"
+	StakeTokenDesc        = "address of staking token"
```
Comment (Stefan-Ethernal) [inline command/secrets/init/utils.go]:
```suggestion
	StakeTokenDesc        = "stake token address"
```

File: command/validator/stake/params.go [hunk 9]
```diff
@@ -30,6 +36,16 @@ func (sp *stakeParams) validateFlags() (err error) {
...
 
+	if err := types.IsValidAddress(sp.stakeToken); err != nil {
+		return fmt.Errorf("stake token address is not a valid address: %w", err)
+	}
+
+	sp.stakeTokenAddr = types.StringToAddress(sp.stakeToken)
+
+	if sp.stakeTokenAddr == types.ZeroAddress {
+		return errStakeTokenIsZeroAddress
+	}
```
Comment (Stefan-Ethernal) [inline command/validator/stake/params.go:47]: We have this code on the multiple places. Probably it is worth considering to expand `IsValidAddress` to return `types.Address` in case address is valid and also to check for zero address (however we should specify if `0x0` address is valid or not via parameter). However we can do that through a separate PR.

Comment (dusannosovic-ethernal) [reply command/validator/stake/params.go:47]: Ok, we will fix it with separate PR.

**Review by Stefan-Ethernal — APPROVED** [review]

**Review by goran-ethernal — APPROVED** [review]
